### PR TITLE
feat: add optional numbering of figures and tables when generating md…

### PR DESCRIPTION
**Pull request by deenine**
… ssp using jinja

Signed-off-by: Doug Chivers <doug.chivers@uk.ibm.com>

## Types of changes

- [ ] Hot fix (emergency fix and release)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation (change which affects the documentation site)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [ ] My code follows the code style of this project.
- [ ] Documentation for my change is up to date?
- [ ] My PR meets testing requirements.
- [ ] All new and existing tests passed.
- [ ] All commits are signed-off.

## Summary

Add numbering of figures and tables to markdown output, number incrementally from the start of the document (not per the list template, where figures are numbered relative to the section, because section numbers are not known at this point). This uses the Pandoc method of identifying captions - image captions are sourced from the image alt text and table captions are sourced from a Table: line above or below the table. 

For example, 

`alt text here`

becomes:

`Figure n - alt text here`

which is rendered by Pandoc to 

_Figure n - alt text here_

## Key links:

- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

### CODE REVIEW

File: trestle/core/commands/author/jinja.py [hunk 1]
```diff
@@ -158,10 +166,33 @@ def jinja_ify(
...
+            else:
+                output_file.open('w', encoding=const.FILE_ENCODING).write(output)
 
         except Exception as e:  # pragma: no cover
             logger.error(f'Unknown exception {str(e)} occured.')
             logger.debug(traceback.format_exc())
             return 1
         return 0
+    
+def _number_captions(md_body) -> str:
```
Comment (enikonovad) [inline trestle/core/commands/author/jinja.py:180]: Please add argument type here

File: trestle/core/commands/author/jinja.py [hunk 2]
```diff
@@ -158,10 +166,33 @@ def jinja_ify(
...
+    """Incrementally number tables and image captions"""
```
Comment (enikonovad) [inline trestle/core/commands/author/jinja.py:181]: Missing dot at the end of the line

File: trestle/core/commands/author/jinja.py [hunk 3]
```diff
@@ -158,10 +166,33 @@ def jinja_ify(
...
+    images = {}
+    tables = {}
+    output = md_body.splitlines()
+    
+    for n, line in enumerate(output):
+        if re.match('^!\[.+\]\(.+\)', line):
```
Comment (enikonovad) [inline trestle/core/commands/author/jinja.py:187]: Can the image be not at the beginning of the line?

**Review by enikonovad — COMMENTED** [review]
Overall looks good to me, but please fix the lint issues by running `make code-lint`

File: trestle/core/commands/author/jinja.py (hunk 2 above)
Comment (deenine) [reply trestle/core/commands/author/jinja.py]: done

File: trestle/core/commands/author/jinja.py (hunk 3 above)
Comment (deenine) [reply trestle/core/commands/author/jinja.py]: it probably always should be at the start of a line, but lets not punish authors who write badly formatted markdown

**Review by enikonovad — APPROVED** [review]
LGTM

**Pull request merged**